pkg/workflows/task.go

```
@@ -0,0 +1,178 @@
+	"io"
+
+	"github.com/pborman/uuid"
+	"github.com/sirupsen/logrus"
+
+	"github.com/supergiant/supergiant/pkg/storage"
+	"github.com/supergiant/supergiant/pkg/workflows/steps"
+)
+
+// Task is a workflow that runs and tracks its progress.
+// A workflow is like a program, while a task is like a process,
+// in terms of an operating system.
+type Task struct {
+	ID           string        `json:"id"`
+	Type         string        `json:"type"`
+	Config       *steps.Config `json:"config"`
+	Status       steps.Status  `json:"status"`
+	StepStatuses []StepStatus  `json:"steps"`
+
+	workflow   Workflow
+	repository storage.Interface
+}
+
+func NewTask(taskType string, repository storage.Interface) (*Task, error) {
+	switch taskType {
+	case DigitalOceanMaster:
+		return newTask(DigitalOceanMaster, GetWorkflow(DigitalOceanMaster), repository), nil
+	case DigitalOceanNode:
+		return newTask(DigitalOceanNode, GetWorkflow(DigitalOceanNode), repository), nil
+	default:
+		w := GetWorkflow(taskType)
+
+		if w != nil {
+			return newTask(taskType, w, repository), nil
+		}
+	}
+
+	return nil, ErrUnknownProviderWorkflowType
+}
+
+func newTask(workflowType string, workflow Workflow, repository storage.Interface) *Task {
+	id := uuid.New()
+
+	return &Task{
+		ID:     id,
+		Type:   workflowType,
+		Status: steps.StatusTodo,
+
+		workflow:   workflow,
+		repository: repository,
+	}
+}
+
+// Run executes all steps of workflow and tracks the progress in persistent storage
+func (w *Task) Run(ctx context.Context, config steps.Config, out io.Writer) chan error {
+	errChan := make(chan error, 1)
+
+	go func() {
+		if w == nil {
+			return
+		}
+
+		// Create list of statuses to track
+		for _, step := range w.workflow {
+			w.StepStatuses = append(w.StepStatuses, StepStatus{
+				Status:   steps.StatusTodo,
+				StepName: step.Name(),
+				ErrMsg:   "",
+			})
+		}
+
+		// Set config to the task
+		w.Config = &config
+		// Save task state before first step
+		w.sync(ctx)
```

Why do you ignore an error?